### PR TITLE
HDDS-4123. Integrate OM Open Key Cleanup Service Into Existing Code

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

This pull request completes the open key cleanup service outlined in the parent Jira HDDS-4120. It implements the `OpenKeyCleanupService` class, and starts and stops the service in `KeyManagerImpl`. The following configurations have been defined to specify the service's behavior:

1. ozone.open.key.cleanup.service.interval (default value 24 hours)
2. ozone.open.key.expire.threshold (default value 1 week)
3. ozone.open.key.cleanup.limit.per.task (default value 1000 keys)

See `ozone-defaults.xml` for their corresponding descriptions. Configurations for service interval and expiration threshold were previously defined in `OzoneConfigKeys` and done as raw numbers. This feature moves them to `OMConfigKeys` and implements them as `TimeDurations`, which requires an update to each spot in the code previously referring to these config values (mostly in test setups).

## What is the link to the Apache JIRA

HDDS-4123

## How was this patch tested?

Integration test has been added.

### CODE REVIEW

**Review by bharatviswa504 — COMMENTED** [review]
Expiry threshold is set to 2 seconds, do we need this to be set in all tests?
Previously this config is not used, but now this will affect, but as service interval is 24 hours it is fine I think so.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java [hunk 1]
```diff
@@ -87,6 +87,21 @@ private OMConfigKeys() {
       "ozone.key.deleting.limit.per.task";
   public static final int OZONE_KEY_DELETING_LIMIT_PER_TASK_DEFAULT = 1000;
 
+  public static final String OZONE_OPEN_KEY_CLEANUP_SERVICE_INTERVAL =
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java:90]: Yes, I added the `ozone.om` prefix.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java [hunk 2]
```diff
@@ -88,7 +89,7 @@ public void init() throws Exception {
     scmId = UUID.randomUUID().toString();
     omId = UUID.randomUUID().toString();
     conf.setBoolean(OZONE_ACL_ENABLED, true);
-    conf.setInt(OZONE_OPEN_KEY_EXPIRE_THRESHOLD_SECONDS, 2);
+    conf.setTimeDuration(OZONE_OPEN_KEY_EXPIRE_THRESHOLD, 2, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMDbCheckpointServlet.java]: We don't need this setting in this or the other tests, I have removed these left over open key expiration configs that were present in tests before this feature.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmAcls.java [hunk 3]
```diff
@@ -80,7 +82,7 @@ public static void init() throws Exception {
     scmId = UUID.randomUUID().toString();
     omId = UUID.randomUUID().toString();
     conf.setBoolean(OZONE_ACL_ENABLED, true);
-    conf.setInt(OZONE_OPEN_KEY_EXPIRE_THRESHOLD_SECONDS, 2);
+    conf.setTimeDuration(OZONE_OPEN_KEY_EXPIRE_THRESHOLD, 2, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmAcls.java]: Does not need to be set, I have removed it per the comment above.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java [hunk 4]
```diff
@@ -1,88 +1,251 @@
...
- * success for keys, then clean up those keys.
+ * Background service to move keys whose creation time is past a given
+ * threshold from the open key table to the deleted table, where they will
+ * later be purged by the {@link KeyDeletingService}.
  */
 public class OpenKeyCleanupService extends BackgroundService {
-
   private static final Logger LOG =
-      LoggerFactory.getLogger(OpenKeyCleanupService.class);
+      LoggerFactory.getLogger(KeyDeletingService.class);
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java]: Fixed, thanks!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java [hunk 5]
```diff
@@ -1,88 +1,251 @@
...
 
-  private final static int OPEN_KEY_DELETING_CORE_POOL_SIZE = 2;
+  // Use only a single thread for open key deletion. Multiple threads would read
+  // from the same table and can send deletion requests for same key multiple
+  // times.
+  private final static int KEY_DELETING_CORE_POOL_SIZE = 1;
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java]: Fixed, thanks!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java [hunk 6]
```diff
@@ -1,88 +1,251 @@
...
     return queue;
   }
 
-  private class OpenKeyDeletingTask implements BackgroundTask {
+  private boolean shouldRun() {
+    if (ozoneManager == null) {
+      // OzoneManager can be null for testing
+      return true;
+    }
+    return ozoneManager.isLeader();
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java]: I am now using the isLeaderReady API here. Let me know if I have done so correctly.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java [hunk 7]
```diff
@@ -1,88 +1,251 @@
...
+
       try {
-        // The new API for deleting expired open keys in OM HA will differ
-        // significantly from the old implementation.
-        // The old implementation has been removed so the code compiles.
-        keyManager.getExpiredOpenKeys(0);
-      } catch (IOException e) {
-        LOG.error("Unable to get hanging open keys, retry in"
-            + " next interval", e);
+        ozoneManager.getOmServerProtocol().submitRequest(null, omRequest);
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java]: Updated the code to use the new API. Let me know if I have done so correctly.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java [hunk 8]
```diff
@@ -1,88 +1,251 @@
...
+
+    Pair<String, String> volumeBucketPair = Pair.of(split[1], split[2]);
+    if (!openKeysPerBucket.containsKey(volumeBucketPair)) {
+      openKeysPerBucket.put(volumeBucketPair, new ArrayList<>());
+    }
+
+    try {
+      OpenKey openKey = OpenKey.newBuilder()
+          .setName(split[3])
+          .setClientID(Long.parseLong(split[4]))
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OpenKeyCleanupService.java]: Yes, there were some bugs here related to separators, good catch. I believe they should be fixed now, and the tests in `TestOpenKeyCleanupService` now use some messy keys with extra separators to make sure they are handled correctly.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOpenKeyCleanupService.java [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+    // Set open key cleanup configurations.
+    conf.setTimeDuration(OMConfigKeys.OZONE_OPEN_KEY_CLEANUP_SERVICE_INTERVAL,
+        openKeyCleanupServiceInterval.getDuration(),
+        openKeyCleanupServiceInterval.getUnit());
+    conf.setTimeDuration(OMConfigKeys.OZONE_OPEN_KEY_EXPIRE_THRESHOLD,
+        openKeyExpireThreshold.getDuration(), openKeyExpireThreshold.getUnit());
+    conf.setInt(OMConfigKeys.OZONE_OPEN_KEY_CLEANUP_LIMIT_PER_TASK,
+        TESTING_TASK_LIMIT);
+
+    cluster = MiniOzoneCluster.newBuilder(conf).build();
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOpenKeyCleanupService.java]: `TestOpenKeyCleanupService` is now parameterized for HA and non-HA testing.

**Review by hanishakoneru — COMMENTED** [review]
Thanks @errose28 for working on this. PR LGTM overall. Minor comments posted inline.

**Comment by elek** [comment]
/pending review comments from Hanisha are not addressed (as fat as I see)

**Comment by kaijchen** [comment]
Hi @errose28 and @captainzmc, i will look into this. Any advice would be appreciated.

**Comment by captainzmc** [comment]
> Hi @errose28 and @captainzmc, i will look into this. Any advice would be appreciated.

Thanks @kaijchen working for this. I think you can rebase to the latest master using the development branch of @errose28. Then resubmit the PR.

**Comment by errose28** [comment]
Thanks for picking this up @kaijchen. Feel free to reach out to me on slack with any questions.

**Pull request closed without merging**